src/Symfony/Component/DependencyInjection/Container.php

```diff
@@ -176,6 +177,14 @@ public function set($id, $service)
         if (null === $service) {
             unset($this->services[$id]);
         }
+
+        if (isset($this->privates[$id])) {
+            if (null === $service) {
+                @trigger_error(sprintf('Unsetting the "%s" private service from outside of the container is deprecated since Symfony 3.2 and won\'t be supported anymore in Symfony 4.0. An exception will be thrown instead in Symfony 4.0.', $id), E_USER_DEPRECATED);
+            } else {
+                @trigger_error(sprintf('Resetting the "%s" private service from outside of the container is deprecated since Symfony 3.2 and won\'t be supported anymore in Symfony 4.0. An exception will be thrown instead in Symfony 4.0.', $id), E_USER_DEPRECATED);
+            }
```

Comment:
I'm wondering if we should not have two different Container classes in Symfony 4.0. Something like a `SimpleServiceContainer` class that doesn't support private services and that justs allows to register and resolve definitions. Then the real `Container` class that allows to support private service and dumped containers when they're built from a `ContainerBuilder`.